### PR TITLE
Kuttl test for best practices Kyverno policies

### PR DESCRIPTION
**Pull request by nsathyaseelan**
- Adding Sample KUTTL tests for best practices policies and refactored the kyverno test resource yaml

### CODE REVIEW

File: .github/workflows/gh-actios.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+name: Kuttl tests
```
Comment (anusha94) [inline .github/workflows/gh-actios.yaml:1]: rename the file to `e2e.yaml`

File: .github/workflows/gh-actios.yaml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  
+jobs:
+  run-e2etest:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Prepare environment
+        run: make kind-create-cluster 
+      - name: Install kyverno Operator
```
Comment (anusha94) [inline .github/workflows/gh-actios.yaml:23]: why do we need this step?

**Review by anusha94 — CHANGES_REQUESTED** [review]
Didn't we decide **against** a separate folder for `tests`? Include everything inside the policy folder itself.

File: best-practices/add_network_policy/kuttl-test/01-policy.yaml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+apiVersion: kuttl.dev/v1beta1
+kind: TestStep
+apply:
+- ../add_network_policy.yaml
+assert:
+- policy-assert.yaml
+unitTest: false
```
Comment (pns-nirmata) [inline best-practices/add_network_policy/kuttl-test/01-policy.yaml:7]: Empty last line is recommended, though not urgent. This red mark above in a few other files too. (https://stackoverflow.com/questions/5813311/whats-the-significance-of-the-no-newline-at-end-of-file-log).

**Review by pns-nirmata — APPROVED** [review]
Looking a good start. Minor comment on newline.

File: .github/workflows/gh-actios.yaml (hunk 2 above)
Comment (pns-nirmata) [reply .github/workflows/gh-actios.yaml]: Yes, should be ok not to have kyverno operator, we are testing kyverno only.

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+          # The target branch of a pull request or the branch/tag of a push
+          ref: ${{ github.base_ref || github.ref_name }}
+
+      - name: Set up Go 
+        uses: actions/setup-go@6edd4406fa81c3da01a34fa6f6343087c207a568
+        with:
+          go-version: 1.18
+
+      - name: Test Policy
+        run: go run ./cmd/cli/kubectl-kyverno test ../kyverno-policies
```
Comment (anusha94) [inline .github/workflows/ci.yaml:36]: Define a make target for this and invoke the target in the pipeline

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+          path: kyverno
+          # The target branch of a pull request or the branch/tag of a push
+          ref: ${{ github.base_ref || github.ref_name }}
+
+      - name: Set up Go 
+        uses: actions/setup-go@6edd4406fa81c3da01a34fa6f6343087c207a568
+        with:
+          go-version: 1.18
+
+      - name: Test Policy
```
Comment (anusha94) [inline .github/workflows/ci.yaml:35]:
```suggestion
      - name: Run Kyverno CLI tests
```

File: .github/workflows/ci.yaml [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@ac593985615ec2ede58e132d2e21d2b1cbd6127c
+        with:
+          path: kyverno-policies
+
+      - name: Clone Kyverno
```
Comment (anusha94) [inline .github/workflows/ci.yaml:22]:
```suggestion
      - name: Clone Enterprise Kyverno
```

File: .github/workflows/ci.yaml [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+on:
+  push:
+    branches:
+      - main
+      - release*
+  pull_request:
+    branches:
+      - main
+      - release*
```
Comment (anusha94) [inline .github/workflows/ci.yaml:11]: I don't think we have a release branch

File: .github/workflows/e2e.yaml [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+  
+jobs:
+  run-e2etest:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Prepare environment
+        run: make kind-create-cluster 
+      - name: Install kyverno Operator
```
Comment (anusha94) [inline .github/workflows/e2e.yaml:23]: why do we need kyverno operator?

File: .github/workflows/e2e.yaml [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+        run: make kind-deploy-kyverno-operator
+      - name: Install kyverno
+        run: make kind-deploy-kyverno
+      - name: Wait for Kyverno to start
```
Comment (anusha94) [inline .github/workflows/e2e.yaml:27]: Add a step to verify deployment instead of sleep

File: Makefile [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+.PHONY: test-kuttl
+test-kuttl: $(KUTTL) ## Run kuttl tests
+	@echo Running kuttl tests... >&2
+	@$(KUTTL) test --config kuttl-test.yaml
+
+## Create kind cluster
+.PHONY: kind-create-cluster
+kind-create-cluster: $(KIND) 
+	@echo Create kind cluster... >&2
+	@$(KIND) create cluster --name $(KIND_NAME)
```
Comment (anusha94) [inline Makefile:50]: This creates a cluster with 1 control plane node. Is that enough for our testing?

File: Makefile [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+kind-deploy-kyverno-operator: $(HELM)
+	@echo Install kyverno chart... >&2
+	@$(HELM) repo add nirmata https://nirmata.github.io/kyverno-charts 
+	@$(HELM) install kyverno-operator --namespace nirmata-kyverno-operator --create-namespace nirmata/kyverno-operator --set imagePullSecret.create=false
+
+.PHONY: kind-deploy-kyverno
+kind-deploy-kyverno: $(HELM) 
+	@echo Install kyverno chart... >&2
+	@$(HELM) repo add nirmata https://nirmata.github.io/kyverno-charts
+	@$(HELM) install kyverno --namespace kyverno --create-namespace nirmata/kyverno --set licenseManager.licenseKey=c2yPkGGVtnbubN8EozSDq8ioQQ7wSffSf6DLV+mv2A794EHC3aHrGmRlZe4MBoC0FHw4x17Wyewgezjy+ldmpLzlxusJFEUqL6dPpJbClVPPWevJGWWJBIaOtCFCbiQfxHxmAPMfY8h+Xuwd629OAK1AzOdzdIljWIIYFPwyysnJTvuP3tw6dFtUCA1Cd34/iTanxjqvHPk/7WGwc67vkv0CWRJANdzT+xRSid9g1mqjrY8OIgp0HsQzIiHpDUKk7aA4QpUQxhkS5Jx5ckuPwpnUtKNed93M7g+Rz9aHJ1g=
```
Comment (anusha94) [inline Makefile:67]: where do you get this key from? And please store this as a secret

File: best-practices/add_network_policy/kuttl-test/02-install-resource.yaml [hunk 12]
```diff
@@ -0,0 +1,7 @@
+apiVersion: kuttl.dev/v1beta1
```
Comment (anusha94) [inline best-practices/add_network_policy/kuttl-test/02-install-resource.yaml:1]: You can rename the directory to `e2e` from `kuttl-test`. Kuttl is only a way to do e2e tests.

**Review by anusha94 — CHANGES_REQUESTED** [review]

**Comment by anusha94** [comment]
@nsathyaseelan One suggestion I have is to raise a separate PR for kyverno cli test. It need not be part of kuttl tests PR.

**Comment by nsathyaseelan** [comment]
Closing this PR as this was implemented in https://github.com/nirmata/kyverno-policies/pull/45

**Pull request closed without merging**